enterprise/frontend/src/metabase-enterprise/whitelabel/components/IllustrationWidget/IllustrationWidget.tsx

```diff
@@ -0,0 +1,214 @@
+  EnterpriseSettings,
+  IllustrationSettingValue,
+} from "metabase-enterprise/settings/types";
+
+import { LighthouseImage, PreviewImage } from "./IllustrationWidget.styled";
+
+export interface StringSetting {
+  value: IllustrationSettingValue | null;
+  default: IllustrationSettingValue;
+}
+
+type IllustrationWidgetProps = {
+  id?: string;
+  setting: StringSetting;
+  onChange: (value: string) => void;
+  onChangeSetting: (key: EnterpriseSettingKey, value: unknown) => void;
+  settingValues: Partial<EnterpriseSettings>;
+  defaultIllustrationLabel: string;
+  customIllustrationSetting:
+    | "login-page-illustration-custom"
+    | "landing-page-illustration-custom"
+    | "no-question-results-illustration-custom"
+    | "no-search-results-illustration-custom";
+  errorMessageContainerId: string;
+};
+
+const MB = 1024 * 1024;
+const IMAGE_SIZE_LIMIT = 2 * MB;
+
+export function IllustrationWidget({
+  id,
+  setting,
+  onChange,
+  onChangeSetting,
+  settingValues,
+  defaultIllustrationLabel,
+  customIllustrationSetting,
+  errorMessageContainerId,
+}: IllustrationWidgetProps) {
+  const value = setting.value ?? setting.default;
+  const [fileName, setFileName] = useState("");
+  const [errorMessage, setErrorMessage] = useState("");
+  const fileInputRef = useRef<HTMLInputElement>(null);
+  const data = useMemo(
+    () => [
+      { label: defaultIllustrationLabel, value: "default" },
+      { label: t`No illustration`, value: "no-illustration" },
+      { label: t`Custom`, value: "custom" },
+    ],
+    [defaultIllustrationLabel],
+  );
+
+  function handleChange(value: IllustrationSettingValue) {
+    setErrorMessage("");
+    // Avoid saving the same value
+    // When setting.value is set to the default value its value would be `null`
+    if (value === (setting.value ?? setting.default)) {
+      return;
+    }
+
+    if (value === "custom" && settingValues[customIllustrationSetting]) {
+      onChange("custom");
+    } else if (value === "custom") {
+      fileInputRef.current?.click();
+    } else {
+      onChange(value);
+    }
+  }
+
+  function handleFileUpload(fileEvent: ChangeEvent<HTMLInputElement>) {
+    if (fileEvent.target.files && fileEvent.target.files.length > 0) {
+      const file = fileEvent.target.files[0];
+      if (file.size > IMAGE_SIZE_LIMIT) {
+        setErrorMessage(
+          t`The image you chose is larger than 2MB. Please choose another one.`,
+        );
+        return;
+      }
+
+      const reader = new FileReader();
+      reader.onload = async readerEvent => {
+        const dataUri = readerEvent.target?.result as string;
+        if (!(await isFileIntact(dataUri))) {
+          setErrorMessage(
+            t`The image you chose is corrupted. Please choose another one.`,
+          );
+          return;
+        }
+        setFileName(file.name);
+        // Setting 2 setting values at the same time could result in one of them not being saved
+        await onChange("custom");
```

Comment:
> In some places you call this with `async`, sometimes without `async`. I think we should make it consistent.

Just for clarification, you mean sometimes I call `onChange` sometimes I call `await onChange` right? If that's the case then the comment above the `await onChange` calls already explains that. Normally I don't want to await things unnecessarily unless I have to, which is why I only have `await` in 2 places where we change 2 setting values in quick succession.

Maybe I could use a better structure like (quick idea)
```ts
function scheduleSettingChanges(settingPromises) {
  // do nothing when there is only one promise, because we don't need to wait when changing only 1 settings

  // await the first setting or both (I'd rather prefer to await only the first one to not waste time unnecessarily)
}
```

usage could look something like this
```ts
scheduleSettingChanges([onChange("no-illustration")])
...
scheduleSettingChanges([onChange("custom"), onChangeSetting(customIllustrationSetting, dataUri)])
```

what do you think?


> Also, we should either use strict typings for onChange, or move "default", "no-illustration", "custom" to be constants.

Totally agree, I think I missed this.

Reply:
Fixed in 83360bfde95995f252df6ff676ec3629ef4fd2cc and 78142c20b0d666e22bd1fb4f2d891e3d0d2c257f


I ended up calling everything with `await` since creating another abstraction isn't worth it, it makes the code too complicated.